docs/configuration/optional-settings.md
# Optional Configuration Settings

## BASE_PATH

Default: None

The base URL path to use when accessing Peering Manager. Do not include the
cheme or domain name. For example, if installed at http://example.com/peering/,
set:

```no-highlight
BASE_PATH = "peering/"
```

---

## DEBUG

Default: `False`

This setting enables debugging. This should be done only during development or
troubleshooting. Never enable debugging on a production system, as it can
expose sensitive data to users (even unauthenticated).

---

## LOGGING

Default: several files will be used for logging in the `logs` directory.

The Django framework on which Peering Manager runs allows for the customization
of logging, e.g. to write logs to file. Please consult the
[Django logging documentation](https://docs.djangoproject.com/en/2.2/topics/logging/)
for more information on configuring this setting.

---

## REDIS

[Redis](https://redis.io/) is an in-memory data store similar to memcached. It
is required to support caching functionality .

Redis is configured using a configuration setting similar to `DATABASE`:

* `HOST` - Name or IP address of the Redis server (use `localhost` if running locally)
* `PORT` - TCP port of the Redis service; leave blank for default port (6379)
* `PASSWORD` - Redis password (if set)
* `CACHE_DATABASE` - Numeric database ID for caching
* `DEFAULT_TIMEOUT` - Connection timeout in seconds
* `SSL` - Use SSL connection to Redis

Example:

```
REDIS = {
    'HOST': 'localhost',
    'PORT': 6379,
    'PASSWORD': '',
    'CACHE_DATABASE': 1,
    'DEFAULT_TIMEOUT': 300,
    'SSL': False,
}
```

---

## CACHE_TIMEOUT

Default: 0

The number of seconds to retain cache entries before automatically invalidating
them. Setting the value to 0 will disable the use of the caching functionality.

---

## CHANGELOG_RETENTION

Default: `90`

The number of days to retain logged changes (object creations, updates, and
deletions). Set this to `0` to retain changes in the database indefinitely.
(Warning: This will greatly increase database size over time having also an
impact on its performances)

---

## LOGIN_REQUIRED

Default: `False`

Setting this to `True` will permit only authenticated users to access Peering
Manager. By default, anonymous users are permitted to access Peering Manager as
read-only.

---

## EMAIL

In order to send email, Peering Manager needs an email server configured. The
following items can be defined within the `EMAIL` setting:

  * `SERVER` - Host name or IP address of the email server (use `localhost` if
    running locally)
  * `PORT` - TCP port to use for the connection (default: 25)
  * `USERNAME` - Username with which to authenticate
  * `PASSSWORD` - Password with which to authenticate
  * `TIMEOUT` - Time to wait for a connection (in seconds)
  * `FROM_ADDRESS` - Sender address for emails sent by Peering Manager
  * `SUBJECT_PREFIX` - Prefix of the subject for outgoing emails

---

## PEERINGDB_USERNAME / PEERINGDB_PASSWORD

These settings are being used to authenticate to PeeringDB.

---

## NAPALM_USERNAME / NAPALM_PASSWORD

Peering Manager will use these credentials when authenticating to remote
devices via the [NAPALM library](https://napalm-automation.net/), if installed.
Both parameters are optional but they are required if you want Peering Manager
to push configurations to your devices.

## NAPALM_ARGS

A dictionary of optional arguments to pass to NAPALM when instantiating a
network driver. See the NAPALM documentation for a
[complete list of optional arguments](http://napalm.readthedocs.io/en/latest/support/#optional-arguments).

## NAPALM_TIMEOUT

Default: `30` seconds

The amount of time (in seconds) to wait for NAPALM to connect to a device.

---

## PAGINATE_COUNT

Default: `20`

Determine how many objects to display per page within each list of objects.

---

## TIME_ZONE

Default: `UTC`

The time zone Peering Manager will for date and time operations.
[List of available time zones](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).

---

## BGPQ3_PATH

Default: `bgpq3`

The path to the BGPQ3 binary. By default, Peering Manager will look for `bgpq3`
in the system `PATH`. An absolute path to the binary is preferred if you need
to change this setting.

## BGPQ3_HOST

Default: `rr.ntt.net`

The host that will be used by BGPQ3 to look for IRR objects.

## BGPQ3_SOURCES

Default: `RIPE,APNIC,AFRINIC,ARIN,NTTCOM,ALTDB,BBOI,BELL,JPIRR,LEVEL3,RADB,RGNET,SAVVIS,TC`

A list of comma separated sources from which we will accept IRR objects.

### BGPQ3_ARGS

Default: `{"ipv6": ["-r", "16", "-R", "48"], "ipv4": ["-r", "8", "-R", "24"]}`

A dictionary with two keys: `ipv6` and `ipv4`. Each value must be a list of
strings to pass to BGPQ3. No spaces are allowed inside strings. If a string has
a space in it, it should be split into two distinct strings.

By default the arguments given will ask BGPQ3 to look for IPv6 prefixes with a
mask length greater than or equal to 16 and less than or equal to 48 and for
IPv4 prefixes with a mask length greater than or equal to 8 and less than or
equal to 24.

---

### NETBOX_API

Default: None

The NetBox API URL to which the queries must be sent to.

### NETBOX_API_TOKEN

Default: None

The API token registered in the NetBox instance to be used in queries.

### NETBOX_DEVICE_ROLES

Default: `["router", "firewall", "switch"]`

The roles that devices must have in the NetBox instance that will be queried.


peering_manager/settings.py
# DO NOT EDIT THIS FILE!
#
# All configuration must be done in the `configuration.py` file.
# This file is part of the Peering Manager code and it will be overwritten with
# every code releases.


import os
import socket

from django.contrib.messages import constants as messages
from django.core.exceptions import ImproperlyConfigured

try:
    from peering_manager import configuration
except ImportError:
    raise ImproperlyConfigured(
        "Configuration file is not present. Please define peering_manager/configuration.py per the documentation."
    )

VERSION = "v1.1.1-dev"
DEFAULT_LOGGING = {
    "version": 1,
    "formatters": {
        "simple": {
            "format": "%(asctime)s | %(levelname)s | %(message)s",
            "datefmt": "%Y-%m-%d %H:%M:%S",
        }
    },
    "handlers": {
        "console": {"class": "logging.StreamHandler", "formatter": "simple"},
        "file": {
            "class": "logging.handlers.TimedRotatingFileHandler",
            "filename": "logs/peering-manager.log",
            "when": "midnight",
            "interval": 1,
            "backupCount": 5,
            "formatter": "simple",
        },
        "peeringdb_file": {
            "class": "logging.handlers.TimedRotatingFileHandler",
            "filename": "logs/peeringdb.log",
            "when": "midnight",
            "interval": 1,
            "backupCount": 5,
            "formatter": "simple",
        },
        "napalm_file": {
            "class": "logging.handlers.TimedRotatingFileHandler",
            "filename": "logs/napalm.log",
            "when": "midnight",
            "interval": 1,
            "backupCount": 5,
            "formatter": "simple",
        },
        "netbox_file": {
            "class": "logging.handlers.TimedRotatingFileHandler",
            "filename": "logs/netbox.log",
            "when": "midnight",
            "interval": 1,
            "backupCount": 5,
            "formatter": "simple",
        },
    },
    "loggers": {
        "peering.manager.peering": {"handlers": ["file"], "level": "DEBUG"},
        "peering.manager.peeringdb": {"handlers": ["peeringdb_file"], "level": "DEBUG"},
        "peering.manager.napalm": {"handlers": ["napalm_file"], "level": "DEBUG"},
        "peering.manager.netbox": {"handlers": ["netbox_file"], "level": "DEBUG"},
    },
}

for setting in ["ALLOWED_HOSTS", "DATABASE", "SECRET_KEY", "MY_ASN"]:
    if not hasattr(configuration, setting):
        raise ImproperlyConfigured(
            "Mandatory setting {} is not in the configuration.py file.".format(setting)
        )

# Set required parameters
ALLOWED_HOSTS = getattr(configuration, "ALLOWED_HOSTS")
DATABASE = getattr(configuration, "DATABASE")
SECRET_KEY = getattr(configuration, "SECRET_KEY")
MY_ASN = getattr(configuration, "MY_ASN")

CSRF_TRUSTED_ORIGINS = ALLOWED_HOSTS

BASE_PATH = getattr(configuration, "BASE_PATH", "")
if BASE_PATH:
    BASE_PATH = BASE_PATH.strip("/") + "/"  # Enforce trailing slash only
DEBUG = getattr(configuration, "DEBUG", False)
LOGGING = getattr(configuration, "LOGGING", DEFAULT_LOGGING)
REDIS = getattr(configuration, "REDIS", {})
CACHE_TIMEOUT = getattr(configuration, "CACHE_TIMEOUT", 0)
CHANGELOG_RETENTION = getattr(configuration, "CHANGELOG_RETENTION", 90)
LOGIN_REQUIRED = getattr(configuration, "LOGIN_REQUIRED", False)
NAPALM_USERNAME = getattr(configuration, "NAPALM_USERNAME", "")
NAPALM_PASSWORD = getattr(configuration, "NAPALM_PASSWORD", "")
NAPALM_TIMEOUT = getattr(configuration, "NAPALM_TIMEOUT", 30)
NAPALM_ARGS = getattr(configuration, "NAPALM_ARGS", {})
PAGINATE_COUNT = getattr(configuration, "PAGINATE_COUNT", 20)
TIME_ZONE = getattr(configuration, "TIME_ZONE", "UTC")
EMAIL = getattr(configuration, "EMAIL", {})
BGPQ3_PATH = getattr(configuration, "BGPQ3_PATH", "bgpq3")
BGPQ3_HOST = getattr(configuration, "BGPQ3_HOST", "rr.ntt.net")
BGPQ3_SOURCES = getattr(
    configuration,
    "BGPQ3_SOURCES",
    "RIPE,APNIC,AFRINIC,ARIN,NTTCOM,ALTDB,BBOI,BELL,JPIRR,LEVEL3,RADB,RGNET,TC",
)
BGPQ3_ARGS = getattr(
    configuration,
    "BGPQ3_ARGS",
    {"ipv6": ["-r", "16", "-R", "48"], "ipv4": ["-r", "8", "-R", "24"]},
)

# Pagination
PER_PAGE_SELECTION = [25, 50, 100, 250, 500, 1000]
if PAGINATE_COUNT not in PER_PAGE_SELECTION:
    PER_PAGE_SELECTION.append(PAGINATE_COUNT)
    PER_PAGE_SELECTION = sorted(PER_PAGE_SELECTION)


# Django filters
FILTERS_NULL_CHOICE_LABEL = "-- None --"
FILTERS_NULL_CHOICE_VALUE = "null"


# Use major.minor as API version
REST_FRAMEWORK_VERSION = VERSION[0:3]
REST_FRAMEWORK = {
    "DEFAULT_VERSION": REST_FRAMEWORK_VERSION,
    "DEFAULT_VERSIONING_CLASS": "rest_framework.versioning.AcceptHeaderVersioning",
    "DEFAULT_AUTHENTICATION_CLASSES": [
        "rest_framework.authentication.SessionAuthentication",
        "peering_manager.api.TokenAuthentication",
    ],
    "DEFAULT_PERMISSION_CLASSES": ["peering_manager.api.TokenPermissions"],
    "DEFAULT_FILTER_BACKENDS": ["django_filters.rest_framework.DjangoFilterBackend"],
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
    "PAGE_SIZE": PAGINATE_COUNT,
}


# Case insensitive search for tags
TAGGIT_CASE_INSENSITIVE = True


# NetBox API configuration
NETBOX_API = getattr(configuration, "NETBOX_API", "")
NETBOX_API_TOKEN = getattr(configuration, "NETBOX_API_TOKEN", "")
NETBOX_DEVICE_ROLES = getattr(
    configuration, "NETBOX_DEVICE_ROLES", ["router", "firewall"]
)

# PeeringDB URLs
PEERINGDB_API = "https://peeringdb.com/api/"
PEERINGDB = "https://peeringdb.com/asn/"
PEERINGDB_USERNAME = getattr(configuration, "PEERINGDB_USERNAME", "")
PEERINGDB_PASSWORD = getattr(configuration, "PEERINGDB_PASSWORD", "")

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


try:
    from peering_manager.ldap_config import *

    LDAP_CONFIGURED = True
except ImportError:
    LDAP_CONFIGURED = False

# If LDAP is configured, load the config
if LDAP_CONFIGURED:
    try:
        import ldap
        import django_auth_ldap

        # Prepend LDAPBackend to the default ModelBackend
        AUTHENTICATION_BACKENDS = [
            "django_auth_ldap.backend.LDAPBackend",
            "django.contrib.auth.backends.ModelBackend",
        ]
    except ImportError:
        raise ImproperlyConfigured(
            "LDAP authentication has been configured, but django-auth-ldap is not installed. You can remove peering_manager/ldap_config.py to disable LDAP."
        )


# Force PostgreSQL to be used as database backend
configuration.DATABASE.update({"ENGINE": "django.db.backends.postgresql"})
# Actually set the database's settings
DATABASES = {"default": configuration.DATABASE}


# Redis
if REDIS:
    REDIS_HOST = REDIS.get("HOST", "localhost")
    REDIS_PORT = REDIS.get("PORT", 6379)
    REDIS_PASSWORD = REDIS.get("PASSWORD", "")
    REDIS_CACHE_DATABASE = REDIS.get("CACHE_DATABASE", 1)
    REDIS_DEFAULT_TIMEOUT = REDIS.get("DEFAULT_TIMEOUT", 300)
    REDIS_SSL = REDIS.get("SSL", False)

    # Caching
    CACHEOPS_ENABLED = CACHE_TIMEOUT > 0
    CACHEOPS_REDIS = "rediss://" if REDIS_SSL else "redis://"
    if REDIS_PASSWORD:
        CACHEOPS_REDIS = "{}:{}@".format(CACHEOPS_REDIS, REDIS_PASSWORD)
    CACHEOPS_REDIS = "{}{}:{}/{}".format(
        CACHEOPS_REDIS, REDIS_HOST, REDIS_PORT, REDIS_CACHE_DATABASE
    )
    CACHEOPS_DEFAULTS = {"timeout": CACHE_TIMEOUT}
    CACHEOPS = {
        "auth.user": {"ops": "get", "timeout": 900},
        "auth.*": {"ops": ("fetch", "get")},
        "auth.permission": {"ops": "all"},
        "peering.*": {"ops": "all"},
        "peeringdb.*": {"ops": "all"},
        "users.*": {"ops": "all"},
        "utils.*": {"ops": "all"},
    }
    CACHEOPS_DEGRADE_ON_FAILURE = True


# Email
if EMAIL:
    EMAIL_HOST = EMAIL.get("SERVER")
    EMAIL_PORT = EMAIL.get("PORT", 25)
    EMAIL_HOST_USER = EMAIL.get("USERNAME")
    EMAIL_HOST_PASSWORD = EMAIL.get("PASSWORD")
    EMAIL_TIMEOUT = EMAIL.get("TIMEOUT", 10)
    SERVER_EMAIL = EMAIL.get("FROM_ADDRESS")
    EMAIL_SUBJECT_PREFIX = EMAIL.get("SUBJECT_PREFIX")


# Application definition
INSTALLED_APPS = [
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "cacheops",
    "debug_toolbar",
    "django_filters",
    "django_tables2",
    "rest_framework",
    "netfields",
    "taggit",
    "taggit_serializer",
    "peering",
    "peeringdb",
    "users",
    "utils",
]

MIDDLEWARE = [
    "debug_toolbar.middleware.DebugToolbarMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "users.middleware.LastSearchMiddleware",
    "utils.middleware.ExceptionCatchingMiddleware",
    "utils.middleware.ObjectChangeMiddleware",
    "utils.middleware.RequireLoginMiddleware",
]

ROOT_URLCONF = "peering_manager.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR + "/templates/"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.template.context_processors.media",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
                "utils.context_processors.settings",
            ]
        },
    }
]

WSGI_APPLICATION = "peering_manager.wsgi.application"


# Password validation
AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"
    },
    {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},
    {"NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"},
]


# Internationalization
LANGUAGE_CODE = "en-us"
USE_I18N = True
USE_L10N = True
USE_TZ = True


# Authentication URL
LOGIN_URL = "/{}login/".format(BASE_PATH)


# Messages
MESSAGE_TAGS = {messages.ERROR: "danger"}


# Static files (CSS, JavaScript, Images)
STATIC_ROOT = BASE_DIR + "/static/"
STATIC_URL = "/{}static/".format(BASE_PATH)
STATICFILES_DIRS = (os.path.join(BASE_DIR, "project-static"),)

# Django debug toolbar
INTERNAL_IPS = ["127.0.0.1", "::1"]

try:
    HOSTNAME = socket.gethostname()
except Exception:
    HOSTNAME = "localhost"
